framework/include/materials/MaterialPropertyStorage.h

```diff
@@ -246,39 +248,41 @@ class MaterialPropertyStorage
    */
   void eraseProperty(const Elem * elem);
 
-  static const std::map<std::string, unsigned int> & propIDs() { return _prop_ids; }
-
-protected:
-  /// Release all internal data structures
-  void releaseProperties();
-
-  /// Internal property storage release helper
-  void releasePropertyMap(HashMap<unsigned int, MaterialProperties> & inner_map);
-
-  // indexing: [element][side]->material_properties
-  std::unique_ptr<PropsType> _props_elem;
-  std::unique_ptr<PropsType> _props_elem_old;
-  std::unique_ptr<PropsType> _props_elem_older;
-
-  /// mapping from property name to property ID
-  /// NOTE: this is static so the property numbering is global within the simulation (not just FEProblemBase - should be useful when we will use material properties from
-  /// one FEPRoblem in another one - if we will ever do it)
-  static std::map<std::string, unsigned int> _prop_ids;
-
   /**
-   * Whether or not we have stateful properties.  This will get automatically
-   * set to true if a stateful property is declared.
+   * @returns The current maximum stored state (0 = none, 1 = old, 2 = older)
    */
-  bool _has_stateful_props;
-
+  unsigned int maxState() const
+  {
+    mooseAssert(_max_state < _storage.size(), "Too big");
+    return _max_state;
+  }
+  /**
+   * @returns The number of stored states (2 = up to old, 3 = up to older)
+   */
+  unsigned int numStates() const { return maxState() + 1; }
+  /**
+   * @returns A range over states to be used in range-based for loops
+   */
+  IntRange<unsigned int> stateIndexRange() const { return IntRange<unsigned int>(0, numStates()); }
   /**
-   * True if any material requires older properties to be computed.  This will get automatically
-   * set to true if a older stateful property is declared.
+   * @returns A range over stateful states to be used in range-based for loops
+   *
+   * Will be an empty range if there are no stateful states
    */
-  bool _has_older_prop;
+  IntRange<unsigned int> statefulIndexRange() const
+  {
+    return hasStatefulProperties() ? IntRange<unsigned int>(1, numStates())
+                                   : IntRange<unsigned int>(0, 0);
+  }
 
-  /// mapping from property ID to property name
-  std::map<unsigned int, std::string> _prop_names;
+protected:
+  using BuildPropertyValuePtr = PropertyValue * (*)();
+
+  /// The actual storage
+  std::array<PropsType, max_state + 1> _storage;
+
+  /// Mapping from stateful property ID to property name
+  std::unordered_map<unsigned int, std::string> _stateful_prop_names;
```

Comment:
Same question here as earlier - why not just `vector`?

Reply:
These are sparse too